index/segment/mem/segment.go

```diff
@@ -22,224 +22,271 @@ package mem
+	termsDict termsDictionary
 
-		writer, reader postings.AtomicID
+	writer struct {
+		sync.Mutex
+		idSet  *idsgen.Map
+		nextID postings.ID
 	}
-
-	// Mapping of field (name and value) to postings list.
-	termsDict termsDictionary
+	readerID postings.AtomicID
 }
 
 // NewSegment returns a new in-memory mutable segment. It will start assigning
-// postings IDs at offset+1.
+// postings IDs at the provided offset.
 func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 	s := &segment{
-		opts:      opts,
-		offset:    int(offset) + 1, // The first ID assigned by the segment is offset + 1.
+		offset:    int(offset),
 		newUUIDFn: opts.NewUUIDFn(),
 		termsDict: newTermsDict(opts),
+		readerID:  postings.NewAtomicID(offset),
 	}
 
 	s.docs.data = make([]doc.Document, opts.InitialCapacity())
 
-	s.ids.writer = postings.NewAtomicID(offset)
-	s.ids.reader = postings.NewAtomicID(offset)
+	s.writer.idSet = idsgen.New(256)
+	s.writer.nextID = offset
 	return s, nil
 }
 
-func (s *segment) Insert(d doc.Document) error {
+func (s *segment) Insert(d doc.Document) ([]byte, error) {
 	s.state.RLock()
 	defer s.state.RUnlock()
 	if s.state.closed {
-		return sgmt.ErrClosed
-	}
-
-	if s.state.sealed {
-		return errSegmentSealed
+		return nil, sgmt.ErrClosed
 	}
 
-	docID, err := d.Validate()
-	if err != nil {
-		return err
-	}
+	{
+		s.writer.Lock()
 
-	if docID != nil {
-		ok, err := s.termsDict.ContainsTerm(doc.IDReservedFieldName, docID)
-		if err != nil {
-			return fmt.Errorf("could not lookup document ID '%s'", docID)
-		}
-		if ok {
-			// Index already contains this document.
-			return nil
-		}
-	} else {
-		uuid, err := s.newUUIDFn()
-		if err != nil {
-			return err
+		b := index.NewBatch([]doc.Document{d})
+		if err := s.prepareDocsWithLocks(b); err != nil {
+			return nil, err
 		}
-		d.Fields = append(d.Fields, doc.Field{
-			Name:  doc.IDReservedFieldName,
-			Value: uuid,
-		})
-	}
 
-	// TODO: Consider supporting concurrent writes by relaxing the requirement that
-	// inserted documents are immediately searchable.
-	s.ids.Lock()
+		// Update the document in case we generated a UUID for it.
+		d = b.Docs[0]
 
-	newID := s.ids.writer.Inc()
-	s.insertDoc(newID, d)
-	err = s.insertTerms(newID, d)
-	s.ids.reader.Inc()
+		s.insertDocWithLocks(d)
+		s.readerID.Inc()
 
-	s.ids.Unlock()
+		s.writer.Unlock()
+	}
 
-	return err
+	return d.ID, nil
 }
 
-func (s *segment) Reader() (index.Reader, error) {
+func (s *segment) InsertBatch(b index.Batch) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
-		return nil, sgmt.ErrClosed
+		return sgmt.ErrClosed
 	}
 
-	maxID := s.ids.reader.Load()
-	r := newReader(s, maxID)
-	return r, nil
-}
+	var err error
+	{
+		s.writer.Lock()
 
-func (s *segment) Seal() error {
-	s.state.Lock()
-	if s.state.sealed {
-		s.state.Unlock()
-		return errSegmentSealed
+		err = s.prepareDocsWithLocks(b)
+		if err != nil && !index.IsBatchPartialError(err) {
```

Comment:
Don't you need to check if partial updates are allowed and insert those still? would be good to add a testcase for this

Reply:
This is actually handled in in `prepareDocsWithLocks`. If we don't support partial updates it will return the error right away, otherwise it will continue validating documents and return a partial error (the corresponding test case is `TestSegmentInsertBatchPartialError `)

Reply:
Cool, good stuff.